woocommerce/woocommerce-ios
# Order Details → Use Basic Background for Products Header

Part of #2281. 

## Design

This changes the “Products” header to use a basic background and a headline font. 

I'll work on removing the “s” if there's only one product in another PR. 

Before | After 
--------|-------
![before](https://user-images.githubusercontent.com/198826/83831610-00b68e00-a6a5-11ea-9736-a321549e3681.png)|![after-light](https://user-images.githubusercontent.com/198826/83831616-03b17e80-a6a5-11ea-8447-33b76acc347a.png)

After Dark | After RTL
--------|-------
![after-dark](https://user-images.githubusercontent.com/198826/83831620-0613d880-a6a5-11ea-96c2-e19ebcc9986e.png)        |       ![after-rtl](https://user-images.githubusercontent.com/198826/83831626-07dd9c00-a6a5-11ea-9687-79c424a35219.png)

After Large Font | After iPad
--------|-------
![after-large-font](https://user-images.githubusercontent.com/198826/83831633-09a75f80-a6a5-11ea-9dc8-9350dca6105a.png)        |       ![after-ipad](https://user-images.githubusercontent.com/198826/83831635-0b712300-a6a5-11ea-8aa8-6039e763e6a6.png)

## Code Changes

### Section Header Component

I could not find a reusable component for this so I created a new reusable section header view named [`PrimarySectionHeaderView`](https://github.com/woocommerce/woocommerce-ios/blob/issue/2281-order-details-white-products-background/WooCommerce/Classes/ViewRelated/ReusableViews/Section%20Headers/PrimarySectionHeaderView.swift). Naming suggestions are welcome. 😄 

### Removing the Separator Line

Removing this line was unexpectedly tricky: 

<img src="https://user-images.githubusercontent.com/198826/83831985-cbf70680-a6a5-11ea-8067-383361004d92.png" width="320">

The acceptable solution I found was manually iterating through the `subviews` of the `UITableViewCell` and hiding it. I created a reusable method for that, named [`findSectionTopSeparator()`](https://github.com/woocommerce/woocommerce-ios/blob/81e40c5bd43de6ea01edeb5167547717f28b4f78/WooCommerce/Classes/Extensions/UITableViewCell%2BHelpers.swift#L21-L49).

### Integrating in OrderDetailsDataSource

To use the new section header view, I added a new `HeaderStyle` property in `OrderDetailsDataSource.Section`:

https://github.com/woocommerce/woocommerce-ios/blob/81e40c5bd43de6ea01edeb5167547717f28b4f78/WooCommerce/Classes/ViewModels/Order%20Details/OrderDetailsDataSource.swift#L878-L897

This felt weird because `Section` is supposed to be a model but `HeaderStyle` is a view definition. I couldn't find a better alternative that keeps this configurable. Please let me know if you can find a better way! 

`OrderDetailsDataSource` is almost at 1000 lines now. 🙈 

## Testing

1. Navigate to an Order.
2. Confirm the Products section has a white background and headline label. 

## Reviewing 

Only 1 reviewer is needed but anyone can review. 

## Submitter Checklist
